### PR TITLE
New mode to validate stm after each training step

### PR DESCRIPTION
**Pull request by Anirk4IR**
The "study_validate" mode was added to allow the evaluation of the stm on a data set.  
After each step of the stm training, the model is called under "study_validate" mode to measure the loss for the same data set that will be used during Recall.

### CODE REVIEW

File: lake/oneshot_cls.py [hunk 1]
```diff
@@ -166,7 +165,7 @@ def main():
           with torch.no_grad():
             for batch_idx_val, (val_data, val_target) in enumerate(val_loader):
 
-              if 0 > MAX_VAL_STEPS > batch_idx_val:
+              if 0 > MAX_VAL_STEPS > batch_idx_val:   ##### y esto pa que!
```
Comment (abdel) [inline lake/oneshot_cls.py:168]: I don't see how this would ever work with positive numbers? 🤷

File: lake/oneshot_cls.py [hunk 2]
```diff
@@ -206,7 +205,7 @@ def main():
   study_loader = torch.utils.data.DataLoader(study_dataset, batch_size=config['study_batch_size'], shuffle=False)
   recall_loader = torch.utils.data.DataLoader(recall_dataset, batch_size=config['study_batch_size'], shuffle=False)
 
-  assert len(study_loader) == len(recall_loader)
+  assert len(study_loader)==len(recall_loader)
```
Comment (abdel) [inline lake/oneshot_cls.py:208]: Can you restore the spacing here between `)==(`? Not sure if it was removed on accident (seems to be the only change.

File: cls_module/cls_module/cls.py [hunk 3]
```diff
@@ -235,7 +243,7 @@ def forward(self, inputs, labels=None, mode=None):  # pylint: disable=arguments-
       softmax_preds = F.softmax(preds, dim=1).argmax(dim=1)
       accuracies[self.ltm_key] = torch.eq(softmax_preds, labels).data.cpu().float().mean()
 
-    if mode in ['study', 'recall']:
+    if mode in ['study','study_validate', 'recall']:
```
Comment (abdel) [inline cls_module/cls_module/cls.py:246]: Minor styling: spacing between comma before `study_validate`

File: lake/oneshot_cls.py (hunk 1 above)
Comment (abdel) [reply lake/oneshot_cls.py:168]: I think the idea behind it is to give the option to limit the number of validation steps. For example, if the validation set has 10,000 samples with a `batch_size` of say `100`, by default you would be doing `100` validation steps in total. 

If you only want to validate on a portion of that (i.e. end it early), you can break the loop by checking against the `MAX_VAL_STEPS`.

I think to make this work it, the condition needs to be: `batch_idx_val >= MAX_VAL_STEPS`?

File: lake/oneshot_cls.py [hunk 4]
```diff
@@ -232,7 +231,15 @@ def main():
...
-      model(study_data, study_target, mode='study')
+      study_train_losses, _ = model(study_data, study_target, mode='study')
+      study_train_loss = study_train_losses['stm']['memory']['loss']
+
+      print('Losses step {}, ite {}: \t PR:{:.6f}\
+      PR mismatch: {:.6f} \t PM-EC: {:.6f}'.format(idx, step,
+                                                      study_train_loss['pr'].item(),
+                                                      study_train_loss['pr_mismatch'].item(),
+                                                      study_train_loss['pm_ec'].item()))
+      model(recall_data, recall_target, mode='study_validate')
```
Comment (abdel) [inline lake/oneshot_cls.py:242]: Have you confirmed that the validation losses get logged in TensorBoard as well?

**Review by abdel — CHANGES_REQUESTED** [review]
Looks good to me overall, just had minor comments and a question :)

File: lake/oneshot_cls.py (hunk 4 above)
Comment (Anirk4IR) [reply lake/oneshot_cls.py:242]: Yes, they are shown in TensorBoard as well.
Screen Shot 2021-07-10 at 2 58 27 pm

File: cls_module/cls_module/cls.py (hunk 3 above)
Comment (Anirk4IR) [reply cls_module/cls_module/cls.py:246]: Noted!

File: lake/oneshot_cls.py (hunk 2 above)
Comment (Anirk4IR) [reply lake/oneshot_cls.py:208]: Noted!

File: lake/oneshot_cls.py (hunk 1 above)
Comment (Anirk4IR) [reply lake/oneshot_cls.py:168]: Agree.
